bson/bsonrw/extjson_parser.go

```diff
@@ -104,6 +104,13 @@ func (ejp *extJSONParser) peekType() (bsontype.Type, error) {
 	case jpsSawEndArray:
 		// this would only be a valid state if we were in array mode, so return end-of-array error
 		err = ErrEOA
+	case jpsSawEndObject:
+		if ejp.peekMode() == jpmObjectMode {
```

Comment:
Looking over this again, I still feel like this is a little fragile. I think I'd prefer not modifying `peekType` if possible.

The comment on L105 justifies why the `peekMode()` is not checked for the array case. So, I thought it would be safe to remove this condition, but that led me astray.

Here is my rough understanding.

`extJSONParser` maintains a stack of modes for parsing nested arrays and objects. `advanceState` calls `pushMode` when encountering an `[` or `{` token, and calls `popMode` when encountering a `]` or `}` token to check that brackets are balanced and correct.

So when `advanceState` returns `jpsSawEndObject`, the mode stack for the object that was being read was just popped. Consequently, I think this condition is only true when popping a nested object. For example:

```
[ { "a" : { "b": 1 } } ]
```

When reading the first `}` this condition will be true.
But when reading the second `}`, we'll pop the `jpmObjectMode`. The mode stack will contain one `jpmArrayMode`. So this condition is false.

In summary, I don't quite understand the need for the check, and I am not sure why removing this check breaks tests.

There are other oddities about `peekType` (e.g. not all states are handled in this `switch`, so I think `peekType` just returns the 0 type in those cases...).

So unless we have a solid understanding, I'd like to vote for another approach. I think we can bypass `peekType` entirely by using `advanceState` (explained below).

Reply:
Great points, and thank you for looking so in-depth into `peekType()`. I think I agree, and I've switched to using `advanceState()` with the algorithm you described below. The only thing we lose is the error-checking in `peekType()`; `advanceState()` advances forward blindly and does not return errors, so now neither does `skipObject()`. There are some disadvantages to this I may bring up offline...

Reply:
Per offline discussion, I am not too concerned with losing some skip validation on malformed BSON. The previous behavior was incorrect on valid BSON.